flocker/testtools/_flaky.py

```diff
@@ -22,10 +43,274 @@ def flaky(jira_key):
+
+
+class _RetryFlaky(testtools.RunTest):
+    """
+    ``RunTest`` implementation that retries tests that fail.
+    """
+    # XXX: This should probably become a part of testtools:
+    # https://bugs.launchpad.net/testtools/+bug/1515933
+
+    def __init__(self, output, run_test_factory, case, *args, **kwargs):
+        self._output = output
+        self._run_test_factory = run_test_factory
+        self._case = case
+        self._args = args
+        self._kwargs = kwargs
+
+    def _run_prepared_result(self, result):
+        """
+        Run the test with a result that conforms to testtools' extended
+        ``TestResult`` interface.
+
+        This overrides a method in base ``RunTest`` which is intended to be
+        overwritten.
+        """
+        flaky = _get_flaky_annotation(self._case)
+        if flaky is not None:
+            return self._run_flaky_test(self._case, result, flaky)
+
+        # No flaky attributes? Then run as normal.
+        return self._run_test(self._case, result)
+
+    def _run_test(self, case, result):
+        """
+        Run ``case`` with the ``RunTest`` we are wrapping.
+
+        :param testtools.TestCase case: The test to run.
+        :param testtools.TestResult result: The test result to report to.
+            Must conform to testtools extended test result interface.
+        :return: The modified ``result``.
+        """
+        run_test = self._run_test_factory(case, *self._args, **self._kwargs)
+        return run_test._run_prepared_result(result)
+
+    def _run_flaky_test(self, case, result, flaky):
+        """
+        Run a test that has been decorated with the `@flaky` decorator.
+
+        :param TestCase case: A ``testtools.TestCase`` to run.
+        :param TestResult result: A ``TestResult`` object that conforms to the
+            testtools extended result interface.
+        :param _FlakyAnnotation flaky: A description of the conditions of
+            flakiness.
+
+        :return: A ``TestResult`` with the result of running the flaky test.
+        """
+        successes = 0
+        results = []
+
+        # Optimization to stop running early if there's no way that we can
+        # reach the minimum number of successes.
+        max_fails = flaky.max_runs - flaky.min_passes
+        while (successes < flaky.min_passes and
+               len(results) - successes <= max_fails):
+            was_successful, result_type, details = self._attempt_test(case)
+            if was_successful:
+                successes += 1
+            results.append((result_type, details))
+        successful = successes >= flaky.min_passes
+
+        flaky_data = flaky.to_dict()
+        flaky_data.update({'runs': len(results), 'passes': successes})
+        flaky_details = {
+            'flaky': text_content(pformat(flaky_data)),
+        }
+        combined_details = _combine_details(
+            [flaky_details] + list(r[1] for r in results))
```

Comment:
:grey_question: nit: forming a list out of a generator looks funny when you could just make a list in place:
`[r[1] for r in results]`. Obviously just a style thing, do so or not at your own discretion.